components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/dao/JDBCSubscriberDAOImpl.java

```diff
@@ -0,0 +1,372 @@
+                findSubscriberExistencePrepStmt.setInt(TENANT_ID, tenantId);
+
+                try (ResultSet rs1 = findSubscriberExistencePrepStmt.executeQuery()) {
+                    return rs1.next();
+                }
+            }
+        } catch (SQLException e) {
+            throw new EntitlementException("Error while checking subscriber existence", e);
+        }
+    }
+
+    private void insertSubscriber(String subscriberId, PublisherDataHolder holder, PublisherPropertyDTO[] propertyDTOs)
+            throws EntitlementException {
+
+        int tenantId = CarbonContext.getThreadLocalCarbonContext().getTenantId();
+
+        Connection connection = IdentityDatabaseUtil.getDBConnection(true);
+        try (NamedPreparedStatement createSubscriberPrepStmt = new NamedPreparedStatement(connection,
+                CREATE_SUBSCRIBER_SQL);
+             NamedPreparedStatement createSubscriberPropertiesPrepStmt = new NamedPreparedStatement(connection,
+                     CREATE_SUBSCRIBER_PROPERTIES_SQL)) {
+            createSubscriberPrepStmt.setString(ENTITLEMENT_MODULE_NAME, holder.getModuleName());
+            createSubscriberPrepStmt.setString(SUBSCRIBER_ID, subscriberId);
+            createSubscriberPrepStmt.setInt(TENANT_ID, tenantId);
+            createSubscriberPrepStmt.executeUpdate();
+
+            for (PublisherPropertyDTO dto : propertyDTOs) {
+                if (dto.getId() != null && StringUtils.isNotBlank(dto.getValue())) {
+
+                    createSubscriberPropertiesPrepStmt.setString(PROPERTY_ID, dto.getId());
+                    createSubscriberPropertiesPrepStmt.setString(DISPLAY_NAME, dto.getDisplayName());
+                    createSubscriberPropertiesPrepStmt.setString(PROPERTY_VALUE, dto.getValue());
+                    createSubscriberPropertiesPrepStmt.setBoolean(IS_REQUIRED, dto.isRequired());
+                    createSubscriberPropertiesPrepStmt.setInt(DISPLAY_ORDER, dto.getDisplayOrder());
+                    createSubscriberPropertiesPrepStmt.setBoolean(IS_SECRET, dto.isSecret());
+                    createSubscriberPropertiesPrepStmt.setString(MODULE, dto.getModule());
+                    createSubscriberPropertiesPrepStmt.setString(SUBSCRIBER_ID, subscriberId);
+                    createSubscriberPropertiesPrepStmt.setInt(TENANT_ID, tenantId);
+
+                    createSubscriberPropertiesPrepStmt.addBatch();
+                }
+            }
+            createSubscriberPropertiesPrepStmt.executeBatch();
+            IdentityDatabaseUtil.commitTransaction(connection);
+        } catch (SQLException e) {
+            IdentityDatabaseUtil.rollbackTransaction(connection);
+            throw new EntitlementException("Error while inserting subscriber details", e);
+        } finally {
+            IdentityDatabaseUtil.closeConnection(connection);
+        }
+    }
+
+    private void updateSubscriber(String subscriberId, PublisherDataHolder holder, PublisherPropertyDTO[] propertyDTOs,
```

Comment:
At glance, it looks like we are having redundant params here.

In high level we need,
subscriberId, moduleName, updatedPropertyDTOs -> when module name changed
or 
subscriberId, updatedPropertyDTOs -> when module name not changed..

We can consider having two methods
`updateSubscriberName(String subscriberId, String moduleName)`
`updateSubscriberProperties(String subscriberId, PublisherPropertyDTO[] updatedPropertyDTOs)`

Reply:
instead updated as `private void updateSubscriber(String subscriberId, String updatedModuleName, PublisherPropertyDTO[] updatedPropertyDTOS, int tenantId)` 
con: there will be a null check for `updatedModuleName`